segment/writer_test.go

```diff
@@ -117,3 +118,116 @@ func TestConcurrentReadersAndWriter(t *testing.T) {
 	require.Greater(t, int(atomic.LoadUint64(&numReads)), 1000)
 	require.Greater(t, int(atomic.LoadUint64(&sealedMaxIndex)), 1000)
 }
+
+func TestWriterRecoversFromWriteFailure(t *testing.T) {
```

Comment:
My understanding is that the following test cases would fail the first `WriteAt` or `Sync`. We could maybe add a test case where some writes pass before a failure happens to simulate a partial write.

Reply:
I actually wrote the tests to fail not the first `WriteAt` initially on the same thought...

But it turns out that because of the way we do buffering and only actually write to the file once per `Append` that there is never more than one `WriteTo` at least at the VFS/file level! See the `flush` method - it's the only place we actually write to the file.

Reply:
Yes I understand that as only one flush is done by batch. The test case that I'm referring to is:
1- Call Append, no error
2- Call Append, error
3- Call Append, no error
Verify that call 1 and 3 are in disk and 2 is not. That would prevent a regression where we rewind a successful write.

Reply:
Maybe I'm still missing something but I think that is what the `non-empty` variants test here - they first append a batch successfully and then fail the next batch.

I pulled that out of the table because I wanted to assert both versions for every other case so it made more sense to just call `t.Run` twice per case. Is that what you mean?

Reply:
🤦 I missed the double t.Run call at the end. Yes that's exactly what I had in mind!